numba/typed/dictobject.py

```diff
@@ -1055,24 +1057,252 @@ def impl_iterator_iternext(context, builder, sig, args, result):
-                d[k] = v
+        if items:
+            for (kt, vt), (k, v) in zip(item_types, items):
+                sig = typing.signature(types.void, dt, kt, vt)
+                args = d, k, v
 
-            context.compile_internal(builder, put, sig, args)
+                def put(d, k, v):
+                    d[k] = v
+
+                context.compile_internal(builder, put, sig, args)
 
     return d
+
+
+# ------------------------------------------------------------------------------
+# Literal dictionaries
+# ------------------------------------------------------------------------------
+
+@intrinsic
+def _mixed_values_to_tuple(tyctx, d):
+    keys = [x for x in d.literal_value.keys()]
+    literal_tys = [x for x in d.literal_value.values()]
+
+    def impl(cgctx, builder, sig, args):
+        lld, = args
+        impl = cgctx.get_function('static_getitem',
+                                  types.none(d, types.literal('dummy')))
+        items = []
+        for k in range(len(keys)):
+            item = impl(builder, (lld, k),)
+            casted = cgctx.cast(builder, item, literal_tys[k], d.types[k])
+            items.append(casted)
+            cgctx.nrt.incref(builder, d.types[k], item)
+        ret = cgctx.make_tuple(builder, sig.return_type, items)
+        return ret
+    sig = types.Tuple(d.types)(d)
+    return sig, impl
+
+
+@overload_method(types.LiteralStrKeyDict, 'values')
+def literalstrkeydict_impl_values(d):
+    # This requires faking a values() iterator simply as a tuple, creating a
+    # type specialising iterator would be possible but horrendous and end up
+    # down the "versioned" loop body route.
+    if not isinstance(d, types.LiteralStrKeyDict):
+        return
+
+    def impl(d):
+        return _mixed_values_to_tuple(d)
+    return impl
+
+
+@overload_method(types.LiteralStrKeyDict, 'keys')
+def literalstrkeydict_impl_keys(d):
+    if not isinstance(d, types.LiteralStrKeyDict):
+        return
+    # create a key iterator by specialising a DictType instance with the
+    # literal keys and returning that
+    t = tuple([x.literal_value for x in d.literal_value.keys()])
+
+    def impl(d):
+        d = dict()
+        for x in t:
+            d[x] = 0
+        return d.keys()
+    return impl
+
+
+# have to lower_builtin as this inherits from tuple and literal, both of which
+# provide a match, hence ambiguous before proper resolution gets a chance
+@lower_builtin(operator.eq, types.LiteralStrKeyDict, types.LiteralStrKeyDict)
+def literalstrkeydict_impl_equals(context, builder, sig, args):
+    tu, tv = sig.args
+    u, v = args
+    pred = tu.literal_value == tv.literal_value
+    res = context.get_constant(types.boolean, pred)
+    return impl_ret_untracked(context, builder, sig.return_type, res)
+
+
+@lower_builtin(operator.ne, types.BaseTuple, types.BaseTuple)
+def tuple_ne(context, builder, sig, args):
+    res = builder.not_(literalstrkeydict_impl_equals(context, builder, sig,
```

Comment:
this is never used? the other [tuple_ne](https://github.com/numba/numba/blob/e043ee705d65abd3a066500be4d62b54bdf3586c/numba/cpython/tupleobj.py#L76) is taking precedence.

Reply:
removed in 4f508cd53